flask/app/participants.py

```diff
@@ -33,75 +38,101 @@
     try:
-        int(limit)
+        page = int(request.args.get("page", default=0))
+        if page < 0:  # zero-indexed pages
+            raise ValueError
     except:
-        abort(400, description="Limit must be a valid integer")
-
+        abort(400, description="page must be a non-negative integer")
     try:
-        int(page)
+        limit = request.args.get("limit")
+        if limit is not None:  # unspecified limit means return everything
+            limit = int(limit)
+            if limit <= 0:  # MySQL accepts 0 but that's just a waste of time
+                raise ValueError
     except:
-        abort(400, description="Page must be a valid integer")
-
-    offset = int(page) * int(limit)
-
-    columns = models.Participant.__table__.columns.keys()
-
-    if order_by_col == "family_codename":
-        order_column = getattr(models.Family, order_by_col)
+        abort(400, description="limit must be a positive integer")
+
+    order_by = request.args.get("order_by", type=str)
+    allowed_columns = [
+        "family_codename",
+        "participant_codename",
+        "notes",
+        "participant_type",
+        "sex",
+        "affected",
+        "solved",
+    ]
+    if order_by is None:
+        order = None  # system default, likely participant_id
+    elif order_by == "family_codename":
+        order = models.Family.family_codename
+    elif order_by in allowed_columns:
+        # Since this is an elif clause, we know family_codename is already handled above.
+        # order_by has been restricted to a known list, so we can safely use getattr
+        order = getattr(models.Participant, order_by)
     else:
-        try:
-            order_column = getattr(models.Participant, order_by_col)
-        except AttributeError:
-            abort(400, description=f"Column name must be one of {columns}")
-
-    try:
-        order = getattr(order_column, order_dir)
-    except AttributeError:
-        abort(400, description=f"Column name must be 'asc' or 'desc'")
+        abort(400, description=f"order_by must be one of {allowed_columns}")
+
+    if order:
+        order_dir = request.args.get("order_dir", type=str)
+        if order_dir == "desc":
+            order = order.desc()
+        elif order_dir == "asc":
+            order = order.asc()
+        else:
+            abort(400, description="order_dir must be either 'asc' or 'desc'")
+
+    filters = []
+    family_codename = request.args.get("family_codename", type=str)
+    if family_codename:
+        filters.append(func.instr(models.Family.family_codename, family_codename))
+    participant_codename = request.args.get("participant_codename", type=str)
+    if participant_codename:
+        filters.append(
+            func.instr(models.Participant.participant_codename, participant_codename)
+        )
+    notes = request.args.get("notes", type=str)
+    if notes:
+        filters.append(func.instr(models.Participant.notes, notes))
```

Comment:
We also need to be able to handle 'notes=null'

Reply:
What would that do?

Reply:
Get records where there are no notes associated with the participant. Right now, if you pass notes=null, no participants are returned. BUT, I guess this isn't relevant from a use-case point of view (presumably, we wouldn't need to filter for participants that do not have notes associated with them).